storage/innobase/buf/buf0buf.cc

```diff
@@ -2991,39 +2991,24 @@ buf_page_get_gen(
   return block;
 }
 
-/********************************************************************//**
-This is the general function used to get optimistic access to a database
-page.
-@return TRUE if success */
 TRANSACTIONAL_TARGET
-bool buf_page_optimistic_get(ulint rw_latch, buf_block_t *block,
-                             uint64_t modify_clock, mtr_t *mtr)
+bool buf_page_optimistic_get(rw_lock_type_t rw_latch, buf_block_t *block,
+                             page_id_t id, uint64_t modify_clock, mtr_t *mtr)
 {
   ut_ad(block);
   ut_ad(mtr);
   ut_ad(mtr->is_active());
   ut_ad(rw_latch == RW_S_LATCH || rw_latch == RW_X_LATCH);
 
-  if (have_transactional_memory);
-  else if (UNIV_UNLIKELY(!block->page.frame))
-    return false;
-  else
-  {
-    const auto state= block->page.state();
-    if (UNIV_UNLIKELY(state < buf_page_t::UNFIXED ||
-                      state >= buf_page_t::READ_FIX))
-      return false;
-  }
-
   bool success;
-  const page_id_t id{block->page.id()};
   buf_pool_t::hash_chain &chain= buf_pool.page_hash.cell_get(id.fold());
   bool have_u_not_x= false;
 
   {
     transactional_shared_lock_guard<page_hash_latch> g
       {buf_pool.page_hash.lock_get(chain)};
-    if (UNIV_UNLIKELY(id != block->page.id() || !block->page.frame))
+    if (UNIV_UNLIKELY(!buf_pool.is_uncompressed(block) ||
+                      id != block->page.id() || !block->page.frame))
       return false;
```

Comment:
I tested this with 79e3a658d905ca42e0ec3102e2334dcd182c8ec9 (#3107) to see if the `transactional_lock_guard` works well:
```sh
perf record -e tx-abort ./mtr innodb.innodb_buffer_pool_resize
perf report
```
In that branch, it is OK; we have much more `xabort()` in the buffer pool lookup functions:
```
Samples: 5K of event 'tx-abort', Event count (approx.): 101716
Overhead  Command   Shared Object  Symbol
  99,98%  mariadbd  mariadbd       [.] buf_page_get_low(page_id_t, unsigned long, unsigned long, buf_block_t*, unsigned long, mtr_t*, dberr_t*, bool)
   0,01%  mariadbd  mariadbd       [.] buf_page_create_low(page_id_t, unsigned long, mtr_t*, buf_block_t*)
   0,00%  mariadbd  mariadbd       [.] buf_page_optimistic_get(rw_lock_type_t, buf_block_t*, page_id_t, unsigned long, mtr_t*)
```
In a longer `sysbench` benchmark, there were more `xabort()` in the optimistic function:
```
Samples: 64K of event 'tx-abort', Event count (approx.): 476326
Overhead  Command   Shared Object  Symbol
  83,71%  mariadbd  mariadbd       [.] buf_page_get_low(page_id_t, unsigned long, unsigned long, buf_block_t*, unsigned long, mtr_t*, dberr_t*, bool)
  15,73%  mariadbd  mariadbd       [.] buf_page_optimistic_get(rw_lock_type_t, buf_block_t*, page_id_t, unsigned long, mtr_t*)
   0,33%  mariadbd  mariadbd       [.] TMLockGuard::TMLockGuard(lock_sys_t::hash_table&, page_id_t)
   0,10%  mariadbd  mariadbd       [.] buf_page_create_low(page_id_t, unsigned long, mtr_t*, buf_block_t*)
   0,08%  mariadbd  mariadbd       [.] buf_page_free(fil_space_t*, unsigned int, mtr_t*)
   0,05%  mariadbd  mariadbd       [.] trx_t::commit_low(mtr_t*)
```
For this branch, the traversal of `buf_pool.chunks` could make the memory transaction too big. I reran the same workload on a merge of 4ac8c4c820ebcff3571a2c67acc4fc41510b2d33 and a4389ec605fe2d027ec09850fb736078c250ce3a, and got this:
```
Samples: 18K of event 'tx-abort', Event count (approx.): 278138
Overhead  Command   Shared Object  Symbol
  99,55%  mariadbd  mariadbd       [.] buf_page_get_low(page_id_t, unsigned long, unsigned long, buf_block_t*, unsigned long, mtr_t*, dberr_t*, bool)
   0,20%  mariadbd  mariadbd       [.] buf_page_init_for_read(unsigned long, page_id_t, unsigned long, bool)
   0,13%  mariadbd  mariadbd       [.] buf_page_create_low(page_id_t, unsigned long, mtr_t*, buf_block_t*)
   0,06%  mariadbd  mariadbd       [.] buf_page_optimistic_get(rw_lock_type_t, buf_block_t*, page_id_t, unsigned long, mtr_t*)
   0,04%  mariadbd  mariadbd       [.] trx_t::commit_low(mtr_t*)
   0,02%  mariadbd  mariadbd       [.] TMLockGuard::TMLockGuard(lock_sys_t::hash_table&, page_id_t)
```
As we can see, the does not appear to be any issue with the `buf_page_optimistic_get()` here either. There are fewer samples, though: The server did not complete the `innodb_fast_shutdown=0`. I had to kill the process because it hung due to running out of buffer pool. With #3107 we would abort the buffer pool resize in that scenario.

Reply:
The removed `Block_hint::buffer_fix_block_if_still_valid()` was also essentially performing a `buf_pool.page_hash` lookup, for acquiring a latch on the hash table slice.

One thing that we can do better is to only buffer-fix the block while holding `buf_pool.page_hash`, and try to acquire the `buf_page_t::lock` after we have released the `buf_pool.page_hash`. The buffer-fix will prevent eviction.

Reply:
With the ccd4b967b4890fe235a638317489e8876855a23e simplification of `buf_page_optimistic_get()`, the throughput numbers for my 60-second 64-thread 64×100k row Sysbench `oltp_update_index` were virtually unchanged between that commit and its parent.